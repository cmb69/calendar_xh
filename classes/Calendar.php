<?php

/**
 * Copyright 2021 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

class Calendar
{
    /** @var bool */
    private $weekStartsMonday;

    /**
     * @param bool $weekStartsMonday
     */
    public function __construct($weekStartsMonday)
    {
        $this->weekStartsMonday = $weekStartsMonday;
    }

    /**
     * @param int $year
     * @param int $month
     * @return array<int,array<int,(int|null)>>
     */
    public function getMonthMatrix($year, $month)
    {
        $result = [];
        $row = [];
        $days = (int) date('t', mktime(1, 1, 1, $month, 1, $year));
        for ($day = 1; $day <= $days; $day++) {
            $dayofweek = $this->getDayOfWeek($year, $month, $day);
            if ($day === 1) {
                for ($i = 0; $i < $dayofweek; $i++) {
                    $row[] = null;
                }
            }
            $row[] = $day;
            if ($day === $days) {
                for ($i = $dayofweek + 1; $i < 7; $i++) {
                    $row[] = null;
                }
            }
            if (count($row) === 7 || $day === $days) {
                $result[] = $row;
                $row = [];
            }
        }
        return $result;
    }

    /**
     * @param int $year
     * @param int $month
     * @param int $day
     * @return int
     */
    private function getDayOfWeek($year, $month, $day)
    {
        $dayofweek = (int) date('w', mktime(1, 1, 1, $month, $day, $year));
        if ($this->weekStartsMonday) {
            $dayofweek = ($dayofweek + 6) % 7;
        }
        return $dayofweek;
    }
}
