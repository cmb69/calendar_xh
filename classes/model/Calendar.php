<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar\Model;

use Calendar\Html2Text;

class Calendar
{
    /** @var array<string,Event> */
    private $events;

    /** @param array<string,Event> $events */
    public function __construct(array $events)
    {
        $this->events = $events;
    }

    /** @return array<string,Event> */
    public function events(): array
    {
        return $this->events;
    }

    public function event(string $id): ?Event
    {
        return $this->events[$id] ?? null;
    }

    public function nextEvent(LocalDateTime $now): ?Event
    {
        $nextevent = null;
        $nextldt = null;
        foreach ($this->events as $event) {
            $ldt = $event->after($now);
            if ($ldt === null) {
                continue;
            }
            if ($nextldt === null || $ldt->compare($nextldt) < 0) {
                $nextevent = $event;
                $nextldt = $ldt;
            }
        }
        return $nextevent;
    }

    /** @return list<Event> */
    public function eventsDuring(int $year, int $month): array
    {
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursDuring($year, $month)) {
                $result[] = $event;
            }
        }
        uasort($result, function (Event $a, Event $b) use ($year): int {
            $dt1 = $a->isBirthday() ? $a->start()->withYear($year) : $a->start();
            $dt2 = $b->isBirthday() ? $b->start()->withYear($year) : $b->start();
            return $dt1->compare($dt2);
        });
        return $result;
    }

    /** @return list<Event> */
    public function eventsOn(LocalDateTime $day, bool $daysBetween): array
    {
        assert($day->hour() === 0 && $day->minute() === 0);
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursOn($day, $daysBetween)) {
                $result[] = $event;
            }
        }
        return $result;
    }

    public function delete(string $id): void
    {
        assert(array_key_exists($id, $this->events));
        unset($this->events[$id]);
    }

    public function toICalendarString(Html2Text $converter, string $host): string
    {
        $res = "BEGIN:VCALENDAR\r\n"
            . "PRODID:-//3-magi.net//Calendar_XH//EN\r\n"
            . "VERSION:2.0\r\n";
        foreach ($this->events() as $id => $event) {
            $res .= $event->toICalendarString($id, $converter, $host);
        }
        $res .= "END:VCALENDAR\r\n";
        return $res;
    }
}
