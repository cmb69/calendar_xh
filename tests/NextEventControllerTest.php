<?php

/**
 * Copyright 2021-2023 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

use ApprovalTests\Approvals;
use Calendar\Infra\DateTimeFormatter;
use Calendar\Infra\EventDataService;
use Calendar\Model\Calendar;
use Calendar\Model\CalendarRepo;
use Calendar\Model\Event;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\TestCase;
use Plib\FakeRequest;
use Plib\View;

class NextEventControllerTest extends TestCase
{
    /** @var array<string,string> */
    private $conf;

    /** @var array<string,string> */
    private $lang;

    /** @var CalendarRepo */
    private $calendarRepo;

    /** @var DateTimeFormatter */
    private $dateTimeFormatter;

    /** @var View */
    private $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->conf = XH_includeVar("./config/config.php", "plugin_cf")["calendar"];
        $this->lang = XH_includeVar("./languages/en.php", 'plugin_tx')["calendar"];
        $this->calendarRepo = new CalendarRepo(vfsStream::url("root/"), ".");
        $this->dateTimeFormatter = new DateTimeFormatter($this->lang);
        $this->view = new View("./views/", $this->lang);
    }

    private function sut(): NextEventController
    {
        return new NextEventController($this->conf, $this->calendarRepo, $this->dateTimeFormatter, $this->view);
    }

    public function testRendersNoEvent(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->cmb()]));
        $request = new FakeRequest(["time" => strtotime("1965-04-16T20:38:00+02:00")]);
        $response = $this->sut()->defaultAction($request);
        $this->assertStringContainsString("No further event scheduled.", $response);
    }

    public function testRendersEventBeforeStart(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->intfcb()]));
        $request = new FakeRequest(["time" => strtotime("2025-04-16T20:38:00+00:00")]);
        $response = $this->sut()->defaultAction($request);
        Approvals::verifyHtml($response);
    }

    public function testRendersMultidayEventBeforeStart(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->easter()]));
        $request = new FakeRequest(["time" => strtotime("2025-04-16T20:38:00+00:00")]);
        $response = $this->sut()->defaultAction($request);
        Approvals::verifyHtml($response);
    }

    public function testRendersRunningEvent(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->intfcb()]));
        $request = new FakeRequest(["time" => strtotime("2025-04-16T21:38:00+00:00")]);
        $response = $this->sut()->defaultAction($request);
        Approvals::verifyHtml($response);
    }

    public function testRendersRunningMultidayEvent(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->easter()]));
        $request = new FakeRequest(["time" => strtotime("2025-04-20T20:38:00+00:00")]);
        $response = $this->sut()->defaultAction($request);
        Approvals::verifyHtml($response);
    }

    public function testIssue51(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->cmb()]));
        $request = new FakeRequest(["time" => strtotime("2021-03-23T12:34:00+00:00")]);
        $response = $this->sut()->defaultAction($request);
        Approvals::verifyHtml($response);
    }

    public function testIssue70(): void
    {
        $this->calendarRepo->save(Calendar::fromEvents([$this->cmb()]));
        $request = new FakeRequest(["time" => strtotime("2021-03-25T12:34:00+00:00")]);
        $response = $this->sut()->defaultAction($request);
        Approvals::verifyHtml($response);
    }

    private function cmb(): Event
    {
        return Event::create("1969-03-24", "1969-03-24", "", "", "cmb", "", "", "###", "", "", "");
    }

    private function intfcb(): Event
    {
        return Event::create(
            "2025-04-16",
            "2025-04-16",
            "21:00",
            "22:45",
            "#INTFCB",
            "",
            "",
            "Guiseppe-Meazza-Stadion",
            "",
            "",
            ""
        );
    }

    private function easter(): Event
    {
        return Event::create("2025-04-20", "2025-04-21", "", "", "easter", "", "", "", "", "", "");
    }
}
