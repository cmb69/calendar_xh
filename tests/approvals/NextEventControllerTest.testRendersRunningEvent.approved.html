
<div class="nextevent_date">
  4/16/2025 10:45 pm</div>
<div class="calendar_marquee_outer ">
  <div class="calendar_marquee">
    <div class="nextevent_event">#INTFCB</div>
    <div class="nextevent_date">Event started:</div>
    <div class="nextevent_date">4/16/2025 9:00 pm</div>
    <div class="nextevent_location">Guiseppe-Meazza-Stadion</div>
  </div>
</div>
